Add Remote Execution and Tool Security checklists, CI docs (#89)
galaxy-iuc/standards: docs/best_practices/security.rst

   — use Galaxy credentials and an application's native environment- or file-based interface.
#. :ref:`Verify downloads and preserve transport security <security-downloads>`
   — require TLS verification, prefer immutable sources, and check downloaded content before use.
#. :ref:`Treat browser-rendered output as active content <security-active-content>`
   — raw HTML, SVG, and JavaScript require an explicit trust and containment decision.

Details
-------

.. _security-shell-quoting:

1. Single-quote dynamic values in shell commands
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The ``<command>`` block is rendered into a script and executed by a shell.
Text parameters, dataset paths, output paths, metadata, and collection values
can contain whitespace or shell metacharacters. Place each dynamic value in
single quotes so the shell receives one literal argument:

.. code-block:: xml

    <!-- WRONG: word splitting and shell expansion are possible. -->
    my_tool --input $input --label "$label" --output $output

    <!-- RIGHT: each value is one literal shell argument. -->
    my_tool --input '$input' --label '$label' --output '$output'

> Double quotes are not equivalent to single quotes: the shell still performs

Reviewer: ```suggestion
**Double quotes are not equivalent to single quotes**: the shell still performs
```

3. Treat identifiers and generated filenames as untrusted
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

``element_identifier``, dataset names, metadata values, and labels are
user-influenced strings. They are not safe merely because Galaxy supplied
them. Galaxy generally works to sanitize these values, but the best practice
is still to treat them as untrusted. IUC reviews repeatedly find identifiers
used as symlink names, temporary paths, output prefixes, or fragments of
generated command files.

Two independent controls may be required:

#. **Shell safety:** single-quote the final path whenever it enters a shell
   command.
#. **Filesystem safety:** map the identifier to a deliberately small filename
   alphabet before creating a file, directory, or link.

.. code-block:: xml

    <!-- WRONG: the raw identifier controls a filename and is unquoted. -->
    ln -s '$input' $input.element_identifier

    <!-- RIGHT when the external tool does not need the original identifier. -->
    ln -s '$input' input.fasta

    <!-- RIGHT when the external tool requires a meaningful filename. -->
    #import re
    #set identifier = str($input.element_identifier)

>     #set safe_identifier = re.sub(r'[^\w-]', '_', identifier)

Reviewer: ```suggestion
    #set safe_identifier = re.sub(r'[^\w-_.]', '_', identifier)
```

Author: Claude thought your suggestion was not compilable as-is but did make a change to expand the category. I hope that is okay - let me know if I should revisit this please.

    ln -s '$input' '${safe_identifier}.fasta'

If the external program does not need the original identifier, prefer a fixed
filename. If it does, preserve the original value separately in data or

> metadata. A replacement such as ``re.sub('[^\\w-]', '_', value)`` can cause

Reviewer: ```suggestion
metadata. A replacement such as ``re.sub('[^\\w-_.]', '_', value)`` can cause
```

Author: See related comment above.

- ``.split()``, ``.strip()``, and other method calls inside ``#set`` or
  ``#echo`` expressions;
- joining lists derived from text parameters;
- formatting strings that combine parameters into shell fragments;
- directly accessing an unsanitized or internal representation;
- disabling sanitization and assuming surrounding quotes compensate.

When transforming a text parameter in Cheetah, explicitly convert the
parameter to ``str`` *before* calling methods such as ``split`` or ``strip``,
then quote each resulting shell argument. In
2025, a single IUC change had to add explicit ``str(...)`` conversions across
13 wrappers because Python-expression use bypassed the expected sanitizer.

Do not reduce this rule to "add ``str`` everywhere": the desired result is
that the value is sanitized exactly once for the context where it is consumed.
Values written to JSON or another data-only config may legitimately preserve
characters that would be unsafe in a shell fragment.

.. _security-cheetah-side-effects:

5. Keep Cheetah templating side-effect free
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Galaxy evaluates Cheetah in ``<command>`` and templated ``<configfile>`` blocks

> on the Galaxy server while preparing the job. This happens before the command

Reviewer: This contradicts what `pulsar.rst` says about `<configfile>`s.

Author: Yeah - I reworded the Pulsar side of this. When config files generate a script that run remotely but the script generation itself runs on the server - I think both sides were technically right but I tried to clarify this on the Pulsar side.

eventual command does not contain template-phase behavior. Galaxy's
`user-defined tool security documentation
<https://docs.galaxyproject.org/en/latest/admin/user_defined_tools.html>`__
describes why untrusted XML/Cheetah tools cannot be treated as ordinary
user-authored jobs.

.. _security-data-not-code:

6. Keep user values as data, never generated code
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Avoid inserting user-controlled values into Python, R, JavaScript, ``awk``,
regular-expression replacement programs, or arbitrary shell snippets.
Quoting rules are different for every language and become difficult to reason
about when interpreters are nested.

Prefer:

- a Galaxy ``<configfiles><inputs .../>`` typed JSON document, or a templated
  ``<configfile>`` only when the target format is consumed strictly as data;
- a tabular key/value file parsed by the wrapped application;
- fixed commands selected with a ``select`` parameter;
- a narrow, explicit allowlist when expression support is the purpose of the
  tool.

Do not expose ``eval``, ``exec``, user-provided script bodies, arbitrary
command options, or "run your own command" escape hatches in a general-purpose

> wrapper. If controlled expression evaluation is unavoidable, allowlist names,

Reviewer: ```suggestion
wrapper. If controlled expression evaluation is unavoidable, define allowlists for names,
```

destination before extraction, reject absolute paths and ``..`` traversal, and
consider links and special files. Extract only the members the tool actually
needs.

.. _security-credentials:

8. Keep credentials out of commands, arguments, and logs
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Passwords, API keys, access tokens, and private endpoints must not appear in:

- literal tool XML or test fixtures;
- generated command text or Galaxy job metadata;
- command-line arguments visible in the process list;
- ``echo`` statements, standard output, standard error, or debug logs;
- output datasets or test assertions.

Galaxy 25.1 introduced ``<credentials>`` requirements that inject variables
and secrets into the job environment without making them Cheetah variables.
Use an application's native environment-variable lookup when it does not copy
the value onto the command line. If the application accepts a credential file,
use a runtime helper that reads the environment and creates a private file with
restrictive permissions, then pass only the file path. Do not use an ordinary
templated ``<configfile>`` for a secret: credentials are intentionally not

> Cheetah variables, and ordinary generated config files are not secret storage.

Reviewer: That sounds something we should probably address at the framework level, maybe with an attribute to `<configfile>`?

Author: I wonder if tools are constructing application-specific files where reuse is less easy than it seems or if they are just creating JSON. Not sure - but this language seems largely mirrored in the XSD and I opened an issue for your comment: https://github.com/galaxyproject/galaxy/issues/23191
